### PR TITLE
Add docker plugin

### PR DESCRIPTION
**Pull request by tuxity**
As discussed in #4111, I have created a plugin to allow us to have some docker metrics. 
For now only running containers, healthy and unhealthy container with an alarm when unhealthy containers > 0

### CODE REVIEW

File: conf.d/python.d/dockerd.conf [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#
+#     user: 'username'
+#     pass: 'password'
+#
+# ----------------------------------------------------------------------
+# AUTO-DETECTION JOBS
+# only one of them will run (they have the same name)
+#
+local:
+    url: 'unix://var/run/docker.sock'
```
Comment (ktsaou) [inline conf.d/python.d/dockerd.conf]: I think we can have another one for the case netdata runs in a container itself:

```
local:
    url: 'unix://host/var/run/docker.sock'
```

So both at the same time. Only one will work.

File: python.d/dockerd.chart.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        self.definitions = CHARTS
+
+    def check(self):
+        if not DOCKER_PY:
+            self.error('\'docker\' package is needed to use docker.chart.py')
+            return False
+
+        self.client = docker.DockerClient(base_url=self.configuration.get('url', 'unix://var/run/docker.sock'))
+
+        return True if self.client else False
```
Comment (l2isbad) [inline python.d/dockerd.chart.py]: `self.client` is always `True`. It is just initialization.

File: python.d/dockerd.chart.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    def get_data(self):
+        data = dict()
+        data['running_containers'] = len(self.client.containers.list())
+        data['healthy_containers'] = len(self.client.containers.list(filters={'health': 'healthy'}))
+        data['unhealthy_containers'] = len(self.client.containers.list(filters={'health': 'unhealthy'}))
+
+        return data or None
```
Comment (l2isbad) [inline python.d/dockerd.chart.py]: `self.client.containers.list()` returns a list of `Containers` objects. Does object provide method/attribute about health status?

If so we can call `self.client.containers.list()` only once then.

Comment (tuxity) [reply python.d/dockerd.chart.py]: It doesn't seems possible, we have to call it multiple times https://docker-py.readthedocs.io/en/stable/containers.html

File: conf.d/python.d/dockerd.conf (hunk 1 above)
Comment (tuxity) [reply conf.d/python.d/dockerd.conf]: Hmmm I don't understand, a 2nd 
```
local:
    url: 'unix://host/var/run/docker.sock'
```

?

File: python.d/dockerd.chart.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        self.definitions = CHARTS
+
+    def check(self):
+        if not HAS_DOCKER:
+            self.error('\'docker\' package is needed to use docker.chart.py')
+            return False
+
+        self.client = docker.DockerClient(base_url=self.configuration.get('url', 'unix://var/run/docker.sock'))
+
+        return True
```
Comment (l2isbad) [inline python.d/dockerd.chart.py]: `check` checks only if python docker package is installed.

We need to check if we can to connect and retrieve data.

Comment (tuxity) [reply python.d/dockerd.chart.py]: I can do a `docker.version()` with a `try` to see if it's ok for more safety

File: python.d/dockerd.chart.py (hunk 3 above)
Comment (tuxity) [reply python.d/dockerd.chart.py]: Oh yes we can do that, I'm adding it

File: conf.d/health.d/dockerd.conf [hunk 5]
```diff
@@ -0,0 +1,8 @@
+template: docker_unhealthy_containers
+      on: docker.unhealthy_containers
+   units: unhealthy containers
+   every: 10s
```
Comment (paulfantom) [inline conf.d/health.d/dockerd.conf]: This is too strict. A lot of containerized services need much more time to bootstrap themselves before they become healthy. For example [minio container will need at least 30s](https://github.com/minio/minio/blob/a63bc9254d481a32f2abebd18fccbfa816c5f8bb/Dockerfile#L31)

**Comment by paulfantom** [comment]
@ktsaou while we are on the subject of monitoring healthy containers, does netdata have HTTP endpoint for healthchecks?

**Comment by paulfantom** [comment]
I mean, do netdata by itself have some endpoint to check if it finished bootstraping process (is healthy)? Something that would show that netdata finished initial `discovery` stage and it is ready to collect metrics (fully operational).

I don't mean to check health status of other services.

File: conf.d/health.d/dockerd.conf (hunk 5 above)
Comment (tuxity) [reply conf.d/health.d/dockerd.conf]: Docker `HEALTHCHECK` directive is in charge of setting the health of a container. https://docs.docker.com/engine/reference/builder/#healthcheck

As you can see, there is also a `starting` health status, so an `unhealthy` container went by `starting` status before. 

I don't think it's netdata "problem", it's an issue of the container configuration, no ?

Comment (paulfantom) [reply conf.d/health.d/dockerd.conf]: Of course it is not netdata problem. Problem is that you are setting very strict alert rule which will cause much problems on a lot of systems by sending false-positive alerts.

Comment (tuxity) [reply conf.d/health.d/dockerd.conf]: But, which false-positive ? if docker says it's unhealthy, it's unhealthy, I don't see it as a false positive. 
What do you suggest ?

Comment (paulfantom) [reply conf.d/health.d/dockerd.conf]: I just ran some tests and seems that problem I was referencing is already fixed in docker daemon. Let's go with your alert setting implementation.

File: conf.d/python.d/dockerd.conf (hunk 1 above)
Comment (paulfantom) [reply conf.d/python.d/dockerd.conf]: I also don't understand that. @ktsaou what do you mean? Containerized netdata shouldn't have 'unix://host/var/run/docker.sock` socket, it is mounted as `'unix://var/run/docker.sock`

As per: https://github.com/firehol/netdata/wiki/Install-netdata-with-Docker#run-netdata-with-docker-command

File: python.d/README.md [hunk 6]
```diff
@@ -526,6 +526,24 @@ localhost:
...
+---
+
+# docker
+
+Module monitor docker health metrics.
+
+**Requirement:**
+* `docker` package
+
+### configuration
```
Comment (l2isbad) [inline python.d/README.md]: please add plugin charts info

**Comment by ktsaou** [comment]
so, merge?

File: conf.d/python.d/dockerd.conf (hunk 1 above)
Comment (ktsaou) [reply conf.d/python.d/dockerd.conf]: ok,

python plugin supports jobs. When 2 jobs have the same `name`, the code picks the first that works.

We use this feature for auto-detection. Let's assume nginx files on ubuntu are on `/var/log/www` and on centos on `/var/log/httpd`. We create 2 jobs with the same name, and the plugin will pick the first job that works. The same is true for example when we try to connect to a port. We try `localhost`, `127.0.0.1` and `::1` - only one will actually be used - the first that succeeds will obsolete the others.

Though the instructions I gave you above are wrong. The right config is:

```
local_run:
    name: 'local'
    url: 'unix://var/run/docker.sock'

local_host:
    name: 'local'
    url: 'unix://host/var/run/docker.sock'
```

So, the first of the above that will succeed, will obsolete the other. Only one job for each `name` is used.

Comment (ktsaou) [reply conf.d/python.d/dockerd.conf]: And now I saw that containerized netdata has the socket as `/var/run/docker.sock`, so the above are not needed.
Sorry...

**Pull request merged**